08_project/n0npax/pkg/puppy/store/syncStore.go

```diff
@@ -0,0 +1,62 @@
+package store
+
+import (
+	"fmt"
+
+	puppy "github.com/anz-bank/go-course/08_project/n0npax/pkg/puppy"
+)
+
+// NewSyncStore creates new storer for SyncMap
+func NewSyncStore() *SyncStore {
+	return &SyncStore{}
+}
+
+// CreatePuppy creates puppy
+func (s *SyncStore) CreatePuppy(p *puppy.Puppy) (int, error) {
+	if p.Value < 0 {
+		return -1, puppy.ErrInvalidInput(puppy.InvalidInputMsg)
+	}
+	s.Lock()
+	defer s.Unlock()
+	p.ID = s.total
+	s.total++
+	s.Store(p.ID, *p)
+	return p.ID, nil
+}
+
+// ReadPuppy reads puppy from backend
+func (s *SyncStore) ReadPuppy(id int) (*puppy.Puppy, error) {
+	if puppyData, ok := s.Load(id); ok {
+		p := puppyData.(puppy.Puppy)
+		return &p, nil
+	}
+	return nil, puppy.ErrNotFound(fmt.Sprintf(puppy.PuppyNotFoundMsg, id))
+}
+
+// UpdatePuppy updates puppy
+func (s *SyncStore) UpdatePuppy(id int, p *puppy.Puppy) error {
+	if p.Value < 0 {
+		return puppy.ErrInvalidInput(puppy.InvalidInputMsg)
```

Comment:
This strikes me as a bit strange - I'd provide the error message here and not go with constants, but I can't really argue well why. My usual approach is to create function akin to this:
```go
func Errorf(code ErrCode, format string, args ...interface{}) *Error
```
I usually declare this one Error method and provide code on creation. It's also useful to mention calling context in the message IMO.
So I'd do something like:

```go
return Errorf(InvalidInput, "UpdatePuppy: puppy value (%d) must not be negative", p.Val)  
```

Reply:
I moved error messages to the requested format.
I provided them as const to avoid running `sed` on the source code.
`Errorf` function refactored. `args ...interface{}` looks awesome, Thanks!